### ISSUE TITLE
Add real-time Instance Segmentation

### ISSUE THREAD

**Issue by AlexeyAB** [AlexeyAB/darknet#3048]
Add real-time Instance Segmentation:

* Look at: https://github.com/pjreddie/darknet/blob/master/examples/instance-segmenter.c

* **YOLACT** (You Only Look At CoefficienTs): https://arxiv.org/abs/1904.02689v1 and https://github.com/dbolya/yolact

    1. Add additional **output-prototype-layer** with full-resolution masks (`W x H x K`) with `upsampling_layers=subsampling_layers`, where is `K=32` number of prototype masks

    2. Use `(4+1+classes+K)*anchors` outputs of Detector instead of `(4+1+classes)*anchors`. Then we do NMS or Fast-NMS. Then we do GEMM between **output-prototype-layer**  `[W x H][K]` and **found N-bboxes** `[N][K]` to get `[WxH][N]` masks for N found bboxes, and apply `logistic_activation`.

**During training:** to compute mask loss, we simply take the pixel-wise binary cross entropy between assembled masks M and the ground truth masks. We crop with the ground truth bounding box and divide Lmask by the ground truth bounding box area.

**Test on:**
* MS COCO
* Kitty
* Cityscapes

![image](https://user-images.githubusercontent.com/4096485/56818231-da8eaf80-684f-11e9-99ac-ed6fa2858279.png)

![image](https://user-images.githubusercontent.com/4096485/56818322-0e69d500-6850-11e9-80cd-9640af348fe4.png)

**Comment by LukeAI** [comment]
This is (another) exciting and impressive potential enhancement.
I was wondering, @AlexeyAB - how do you find time to do all this stuff and then give it away for free - are you being paid to produce this stuff or are you a passionate hobbyist?

**Comment by abhigoku10** [comment]
@AlexeyAB  have you implemented the yolAct architecture in the instance segmenter.c ??

**Comment by zpmmehrdad** [comment]
Hi @AlexeyAB,

I really need it. When do you think we can use this feature?

Thanks in advance.

**Comment by LukeAI** [comment]
Why not use https://github.com/dbolya/yolact ?

**Comment by abhigoku10** [comment]
@zpmmehrdad  suggest to use @LukeAI  has suggest its good in fps and is implemented in pytorch

**Comment by zpmmehrdad** [comment]
@abhigoku10 Hi,

Thanks for suggesting, I want to use it in C++.

**Comment by abhigoku10** [comment]
@zpmmehrdad yes even i wanted to use c++ later i shifted my whole development environment pls let me know if come across anyother repo

**Comment by WongKinYiu** [comment]
CenterMask:Real-Time Anchor-Free Instance Segmentation: https://arxiv.org/pdf/1911.06667.pdf
![image](https://user-images.githubusercontent.com/12152972/69897505-f333ea80-1387-11ea-9d91-9000e804097d.png)

Faster and better than YOLACT.
![image](https://user-images.githubusercontent.com/12152972/69897515-12327c80-1388-11ea-92b2-6816b43bc3b3.png)

**Comment by Kyuuki93** [comment]
YOLACT++: Better Real-time Instance Segmentation

* paper https://arxiv.org/abs/1912.06218
* code https://github.com/dbolya/yolact (same with yolact)

| Method| mAP | FPS |
| --- | --- | ---|
|yoloact | 29.8%| 33.0|
|yoloact++|34.1%| 33.5|
